### PR TITLE
Support entitlement enforcement

### PR DESCRIPTION
**Pull request by pratikgadiya12**
Fixes: https://projects.engineering.redhat.com/browse/OSBS-4839
Co-Authored-By: Rajshekhar AndalaPisharam <randalap@redhat.com>

### CODE REVIEW

File: crane/app_util.py [hunk 1]
```diff
@@ -71,9 +71,16 @@ def repo_is_authorized(repo_id):
 
     if repo_tuple.protected:
         cert = _get_certificate()
-        if not cert or not cert.check_path(repo_tuple.url_path):
+        logger.info('cert when checking repo %s' % cert)
```
Comment (twaugh) [inline crane/app_util.py]: Should this (and maybe others) be `logger.debug`?

File: crane/app_util.py [hunk 2]
```diff
@@ -71,9 +71,16 @@ def repo_is_authorized(repo_id):
...
+        value = repo_tuple.url_path
+        logger.info('path before check %s' % value)
+        if 'webassets/docker' in value:
```
Comment (twaugh) [inline crane/app_util.py]: This string is repeated in several places. Should it be stored in a variable at module scope, or perhaps in configuration?

Alternatively, perhaps this logic should factored into a function that can be called from the various places it is used, like: `if not cert or not cert.check_path(remove_prefix(path)): ...`.

File: crane/app_util.py [hunk 3]
```diff
@@ -107,9 +114,16 @@ def wrapper(image_id, *args, **kwargs):
             if not repo_tuple.protected:
                 found_match = True
                 break
-            elif cert and cert.check_path(repo_tuple.url_path):
-                found_match = True
-                break
+            elif repo_tuple.protected:
```
Comment (twaugh) [inline crane/app_util.py]: Looks like this should just be `else:`, or even just a separate block outside of the `if` scope like:
```
if not repo_tuple.protected:
    found_match = True
    break

value = ...
```

File: crane/app_util.py [hunk 4]
```diff
@@ -107,9 +114,16 @@ def wrapper(image_id, *args, **kwargs):
...
+                value = repo_tuple.url_path
+                logger.info('path before check for images %s' % value)
+                if 'webassets/docker' in value:
```
Comment (twaugh) [inline crane/app_util.py]: Using `in` here (and in the other places this code is repeated) looks wrong. Should it be `.startswith()`?

File: tests/data/metadata_good/foo_webassets_v1.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
+{
+    "images": [
+        {
+            "id": "foowebassets123"
+        }
+    ],
+    "repository": "foo-webassets",
+    "repo-registry-id": "foo/webassets",
```
Comment (twaugh) [inline tests/data/metadata_good/foo_webassets_v1.json]: This doesn't look like the sort of repository or repo-registry-id we'd expect to see. Why is 'webassets' there?

File: crane/app_util.py (hunk 1 above)
Comment (pratikgadiya12) [reply crane/app_util.py]: Ideally, this should be logger.debug but since we don't have production environment replica to test this out I have kept it to logger.info so that I could see what has been failed and could correct that.

If we want to keep this as logger.debug, we have to enable debug and restart the crane service in the production environment to track if in case this PR fails to work and if everything goes fine then we have to raise another PR to make log levels back to info and restart the crane service.

I am ok with both the approaches

File: crane/app_util.py (hunk 2 above)
Comment (pratikgadiya12) [reply crane/app_util.py]: Since we are planning to move to Openshift registry, I think this patch is only required for this release and eventually, it would be reverted/removed when we move to Openshift registry.
So, I thought this was an extra effort to make it more modularize

File: crane/app_util.py (hunk 4 above)
Comment (pratikgadiya12) [reply crane/app_util.py]: value contains 'webassets/docker' in between so I have used `in` here

Comment (twaugh) [reply crane/app_util.py]: But the next line strips off a string the length of 'webassets/docker', so if it wasn't at the start won't it remove the wrong characters?

File: crane/app_util.py [hunk 6]
```diff
@@ -107,9 +114,16 @@ def wrapper(image_id, *args, **kwargs):
...
+                    value = value[len('webassets/docker') + 1:]
```
Comment (twaugh) [inline crane/app_util.py]: Should the checked-for prefix be 'webassets/docker/' instead, including the trailing slash? Otherwise it is not clear why 1 is being added to the length here.

Comment (pratikgadiya12) [reply crane/app_util.py]: value generated is ''/webassets/docker/content/dist/rhel/server/7/7Server/x86_64/containers/registry/'' so +1 was added to escape starting "/" from value.

Let me make it value.startswith('/webassets/docker') and value = value[len('/webassets/docker'):]

File: crane/app_util.py (hunk 4 above)
Comment (pratikgadiya12) [reply crane/app_util.py]: value generated is ''/webassets/docker/content/dist/rhel/server/7/7Server/x86_64/containers/registry/'

Let me make this startswith('/webassets/docker')

**Comment by twaugh** [comment]
It is not running in the current production environment. (Is it?) Isn't that the whole point of reviewing this? We have a test environment in place to test this, and a schedule for testing.

**Comment by pratikgadiya12** [comment]
@twaugh: Let me explain this from beginning

Events:
1. Production environment running with upstream crane repository

2. Error occurs, AP fixes the error (almost a year ago). The error gets resolved and now the production environment is patched one and not the upstream crane.

3. From the time when AP had fixed the error and we decide to support new pulp upgrade, new features mentioned below were introduced in upstream pulp/crane repositories
    a. Multi-Arch
    b. Manifest Lists

4. In order to support these new features, we need to port AP's previous patch to that of the current development in crane (which I have done). So, that we could go ahead with our pulp upgrade and don't come across error we faced earlier in the production environment.

I have tried to explain the events from the beginning.
Let me know if I was able to explain it

**Comment by twaugh** [comment]
Ah, I see. Thanks for the explanation.

But yes, I would prefer changes be made to address the comments here before moving on to build + testing. In particular, the discrepancy between `in` and `[len(...):]` should definitely be fixed as it is a bug.

**Comment by pratikgadiya12** [comment]
Thanks for confirming.

I didn't get how `in` and `[len(...):]` is a bug.

Current Code (using in):-
Here `value` is `'/webassets/docker/content/dist/rhel/server/7/7Server/x86_64/containers/registry/'`        

> if 'webassets/docker' in value:
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;value = value[len('webassets/docker') + 1:]

After manipulation `value` becomes `'/content/dist/rhel/server/7/7Server/x86_64/containers/registry/'`

After code changes(using startswith):
Here `value` is same as above i.e. `'/webassets/docker/content/dist/rhel/server/7/7Server/x86_64/containers/registry/'`


> if value.startswith('/webassets/docker'):
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;value = value[len('/webassets/docker'):]

 

After manipulation `value` becomes same as above i.e. `'/content/dist/rhel/server/7/7Server/x86_64/containers/registry/'`

It's just a better coding part, right?

**Comment by twaugh** [comment]
It's a bug because it makes an assumption about the input. "Here `value` is '/webassets/docker/...`" -- but what if it happens to be '/foo/webassets/docker/...' at some point in the future, for some reason we don't yet know? In that case, the string will be entirely messed up.

**Comment by pratikgadiya12** [comment]
Now it's clear. Makes sense

**Comment by ipanova** [comment]
i went through the changes
1) i'd move info to debug but since this will be ran just by you and if you're ok with that i will leave this up to you.
2) as Tim suggested i'd store this check in a separate function to be conform the dry principles since we are checking same thing in 3 places. But if you think this does not worth because we'll soon get rid of this patch, fine with me as well.
3) maybe leave a comment in the code to explain the purpose of string slicing
4) i think multi arch and manifest list are the same thing :) 
5) do we support v1 content protection? i thought we moved to v2

**Comment by pratikgadiya12** [comment]
Thanks Ina for the review.

1. Keeping this to info.
2. We will get soon rid of this patch, so keeping it as it is.
3. Done. Comment added to explain the purpose of string slicing.
4. Yes
5. We did move to V2, however, there are still some repositories in crane production environment which uses content protection so have added the code to v1 as well.

File: crane/app_util.py [hunk 7]
```diff
@@ -71,9 +71,19 @@ def repo_is_authorized(repo_id):
 
     if repo_tuple.protected:
         cert = _get_certificate()
-        if not cert or not cert.check_path(repo_tuple.url_path):
+        logger.info('cert when checking repo %s' % cert)
+        value = repo_tuple.url_path
+        logger.info('path before check %s' % value)
+        # In production environment, due to akamine storage string
```
Comment (twaugh) [inline crane/app_util.py]: Looks like a typo?

File: crane/app_util.py [hunk 8]
```diff
@@ -107,9 +117,19 @@ def wrapper(image_id, *args, **kwargs):
             if not repo_tuple.protected:
                 found_match = True
                 break
-            elif cert and cert.check_path(repo_tuple.url_path):
-                found_match = True
-                break
+            elif repo_tuple.protected:
+                value = repo_tuple.url_path
+                logger.info('path before check for images %s' % value)
+                # In production environment, due to akamine storage string
```
Comment (twaugh) [inline crane/app_util.py]: Typo

File: crane/app_util.py [hunk 9]
```diff
@@ -260,9 +280,19 @@ def name_is_authorized(name):
 
     if v2_repo_tuple.protected:
         cert = _get_certificate()
-        if not cert or not cert.check_path(v2_repo_tuple.url_path):
+        logger.info('cert when checking repo %s' % cert)
+        value = v2_repo_tuple.url_path
+        logger.info('path before check %s' % value)
+        # In production environment, due to akamine storage string
```
Comment (twaugh) [inline crane/app_util.py]: Typo

**Comment by pratikgadiya12** [comment]
Ahha !! Good catch

**Comment by ipanova** [comment]
@pratikgadiya12 i am maybe missing something, but how you plan content protect repositories that contains multi arch? i think there are different entitlements for different arches and the protection comes per repo. Or you just just plan to keep the content protection for older repos that do not contain manifest lists and those which contain manifest lists would be unprotected?

**Comment by pratikgadiya12** [comment]
@ipanova 

tl;dr For both V1 and V2 each and every call gets checked for entitlement cert based on whether the repo is protected or not (as authorization decorators are used on all the methods that process tags, manifests, multi-arch data, images etc..)
In V1, it is for tags, images.
In V2, it is for schema2data and manifest list data

Hence, if the repo is protected everything that gets called from repo is checked(this is as per the code that is written in v1.py and v2.py)

To explain

----------------------------
Looking at the code v2.py inside `name_redirect` method, we have `get_schema2_data_for_repo`, `get_manifest_list_data_for_repo`, `get_manifest_list_amd64_for_repo` methods which are decorated with `authorize_name` decorator, which makes call to `name_is_authorized`. In this `name_is_authorized` method, our patch is present and hence for every call for every check per repo, per list, per tag, etc.. entitlement gets checked and if the url has `'webassets/docker'` then our patch creates a temporary variable, removes the `'webassets/docker'` string and uses the temporary variable to continue the cert check. This was for V2.


Similarly for V1, looking at the code v1.py in `repo_images`, `repo_tags` and `repo_tags_get_tag` methods we have called methods such as `get_images_for_repo`, `get_tags_for_repo` which are decorated with `authorize_repo_id` decorator which calls `repo_is_authorized`. In this `repo_is_authorized`, our patch is present and hence if the url has `'webassets/docker'` then our patch creates a temporary variable, removes the `'webassets/docker'` string and uses the temporary variable to continue the cert check. 

----------------------------

**Comment by ipanova** [comment]
@pratikgadiya12 i think you're good to go, @twaugh if you have any other further comments, i'm cool :)

**Comment by twaugh** [comment]
My only remaining comment is about changing `elif repo_tuple.protected:` to `else:` for clarity only.

I'm fine with the code as it currently stands.
